### PR TITLE
New module to support BGP configuration management in IOS

### PR DESCRIPTION
**Pull request by NilashishC**
##### SUMMARY
- Added new module - `ios_bgp`
- This module provides configuration management of global BGP parameters
    on devices running Cisco IOS
- Targets [this](https://github.com/ansible-network/network-proposals/issues/5).

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
ios_bgp.py

##### ANSIBLE VERSION
```
devel
```

### CODE REVIEW

**Comment by NilashishC** [comment]
One of the sanity checks is expecting `__init__`s to be empty, hence failing for `module_utils/network/ios/bgp/__init__.py` and `bgp/config/__init__.py`. Need to discuss with @privateip regarding this.

**Comment by abenokraitis** [comment]
cc @cidrblock

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+    description:
+      - Specify networks to announce via BGP
+    suboptions:
+      network:
+        description:
+          - Network ID to announce via BGP
+        required: True
+      mask:
+        description:
+          - Subnet Mask for the Network to announce
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
          - Subnet Mask for the network to announce
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+        description:
+          - Specify networks to announce via BGP
+        suboptions:
+          network:
+            description:
+              - Network ID to announce via BGP
+            required: True
+          mask:
+            description:
+              - Subnet Mask for the Network to announce
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
              - Subnet Mask for the network to announce
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+              - present
+              - absent
+      af_neighbors:
+        description:
+          - Specifies Address Family neighbor related configurations
+          - This option extends the top-level parameter - neighbors 
+        suboptions:
+          activate:
+            description:
+              - Enable the Address Family for this Neighbor
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
              - Enable the address family for this neighbor
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+"""
+
+EXAMPLES = """
+- name: configure global bgp as 65000
+  ios_bgp:
+    bgp_as: 65000
+    router_id: 1.1.1.1
+    log_neighbor_changes: True
+    neighbors:
+      - neighbor: 182.168.10.1
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
      - neighbor: 192.168.10.1
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+      - absent
+      - replace
+extends_documentation_fragment: ios
+"""
+
+EXAMPLES = """
+- name: configure global bgp as 65000
+  ios_bgp:
+    bgp_as: 65000
+    router_id: 1.1.1.1
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
    router_id: 10.1.1.1
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+    default: present
+    choices:
+      - present
+      - absent
+      - replace
+extends_documentation_fragment: ios
+"""
+
+EXAMPLES = """
+- name: configure global bgp as 65000
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
- name: configure global bgp as 64496
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+  ios_bgp:
+    bgp_as: 65000
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
    bgp_as: 64496
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    router_id: 1.1.1.1
+    log_neighbor_changes: True
+    neighbors:
+      - neighbor: 182.168.10.1
+        remote_as: 500
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
        remote_as: 64511
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+      - neighbor: 192.168.20.1
+        remote_as: 500
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
        remote_as: 64511
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+    networks:
+      - network: 10.0.0.0/8
+      - network: 11.0.0.0/8
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
      - network: 192.0.2.0/24
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+    state: present
+
+- name: remove bgp as 65000 from config
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
- name: remove bgp as 64496 from config
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+  ios_bgp:
+    bgp_as: 65000
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
    bgp_as: 64496
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+    state: absent
+"""
+
+RETURN = """
+commands:
+  description: The list of configuration mode commands to send to the device
+  returned: always
+  type: list
+  sample:
+    - router bgp 500
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
    - router bgp 64511
```

File: lib/ansible/modules/network/ios/ios_bgp.py [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    - neighbor 182.168.10.1 remote-as 500
```
Comment (samccann) [inline lib/ansible/modules/network/ios/ios_bgp.py]:
```suggestion
    - neighbor 192.168.10.1 remote-as 64511
```

File: test/units/modules/network/ios/fixtures/ios_bgp_config.cfg [hunk 15]
```diff
@@ -0,0 +1,11 @@
+router bgp 65536
+ bgp router-id 192.0.2.1
+ network 11.0.0.0 mask 255.255.0.0 route-map RMAP
```
Comment (samccann) [inline test/units/modules/network/ios/fixtures/ios_bgp_config.cfg]: I don't want to touch test cases, but I worry that we have real IP addresses and as numbers here (not private ones)...

**Review by samccann — COMMENTED** [review]
changing documentation to use doc safe IP addresses and AS numbers.

**Comment by NilashishC** [comment]
> changing documentation to use doc safe IP addresses and AS numbers.

@samccann I have updated all the IPs and ASNs to whats permitted here https://tools.ietf.org/html/rfc5737

**Review by samccann — APPROVED** [review]
docs portion LGTM

**Review by ganeshrn — APPROVED** [review]

**Pull request merged**